### ISSUE TITLE
Undefined reference with virtual workspace

### ISSUE THREAD

**Issue by tobithiel** [rust-ndarray/ndarray-linalg#180]
I created this dummy repository to demonstrate my issue: https://github.com/tobithiel/ndarray_blas_issue

There is library `liba` which depends on `ndarray` but nothing else and library `libb` which depends on `ndarray` and `ndarray-linalg`. Finally there is a binary which depends only on library `liba`

I can compile each project independently, but building the virtual workspace, which includes all of them, yields

```
$ cargo clean && cargo build
   Compiling autocfg v1.0.0
   Compiling libc v0.2.66
   Compiling serde v1.0.104
   Compiling rand_core v0.4.2
   Compiling ndarray v0.13.0
   Compiling rawpointer v0.2.1
   Compiling either v1.5.3
   Compiling blas-src v0.2.1
   Compiling blas-src v0.4.0
   Compiling lapack-src v0.4.0
   Compiling rand_core v0.3.1
   Compiling matrixmultiply v0.2.3
   Compiling itertools v0.8.2
   Compiling num-traits v0.2.11
   Compiling num-complex v0.2.4
   Compiling num-integer v0.1.42
   Compiling rand v0.5.6
   Compiling cblas-sys v0.1.4
   Compiling lapacke-sys v0.1.4
   Compiling lapacke v0.2.0
   Compiling cauchy v0.2.2
   Compiling ndarray-linalg v0.12.0
   Compiling liba v0.1.0 (~/ndarray_blas_issue/lib/liba)
   Compiling bina v0.1.0 (~/ndarray_blas_issue/bin/bina)
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.101g5gvj4752wtsj.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.190adlbyk4vwwpsf.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.1gd4ed2qs138zz72.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.1mo9nby75d368fpo.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.1oeasc9sgs0revlt.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.1q3lp44ohhhil6nu.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.1qyt9mz0gqc1tqrc.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.1tbw2yhs6r1nvjmk.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.25u7kkv3gnhqqq5a.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.2a4cv4ru2fe9p6vv.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.2gzx29ty68nnn8jf.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.2js9f15fvy5yw5ma.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.2lamef3c9jsbd0f.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.2r414auq17d6rcb0.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.2yvdqweb5bathunt.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.32cpukh45l8g23yu.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.3hdtu5ewfahp1ckg.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.3jfu5mxbx3eara4f.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.3l917gg0785e79eq.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.3rrlhg950pb9theu.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.3ujzqtwiuhsz133b.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.43qbruwcay4ave6s.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.4cqg927txu6fwvs3.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.4dhc8pwtvdumh7pe.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.4mhjjuff9rx2w45a.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.4nb4f724r6xu40it.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.4nb4snpdi9hp3j04.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.4v121e00jq5jph2l.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.4xcw4rfpqe36z779.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.53gyl7xqsucllzik.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.5bea5rpe83t6x1yj.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.5bewehlgts0xz9n.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.5gptzh4x4xdluvvc.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.entld3runmcivpe.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.gt16w14tux937uo.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.scozvhrn6kbvw99.rcgu.o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.z2rww29492aygqa.rcgu.o" "-o" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1" "~/ndarray_blas_issue/target/debug/deps/bina-6a3eb0fe8bddc6d1.2g28fkvd9qehb46g.rcgu.o" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-nodefaultlibs" "-L" "~/ndarray_blas_issue/target/debug/deps" "-L" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "~/ndarray_blas_issue/target/debug/deps/libliba-362c1514609b1fc9.rlib" "~/ndarray_blas_issue/target/debug/deps/libndarray-319295d97b726eea.rlib" "~/ndarray_blas_issue/target/debug/deps/libmatrixmultiply-5cda06ee4802c361.rlib" "~/ndarray_blas_issue/target/debug/deps/libnum_complex-a7a46237962d615e.rlib" "~/ndarray_blas_issue/target/debug/deps/librand-0e0265e96be67045.rlib" "~/ndarray_blas_issue/target/debug/deps/librand_core-5c90dfdf043d3307.rlib" "~/ndarray_blas_issue/target/debug/deps/librand_core-94dc1c6006443d6f.rlib" "~/ndarray_blas_issue/target/debug/deps/libserde-852ab9d6e4ec13a2.rlib" "~/ndarray_blas_issue/target/debug/deps/libnum_integer-02a05adc03fcc2e7.rlib" "~/ndarray_blas_issue/target/debug/deps/libitertools-1662921f0e878972.rlib" "~/ndarray_blas_issue/target/debug/deps/libeither-09d857519729919f.rlib" "~/ndarray_blas_issue/target/debug/deps/libnum_traits-3a228a0cbaa6adf6.rlib" "~/ndarray_blas_issue/target/debug/deps/librawpointer-db446963b9e301fb.rlib" "~/ndarray_blas_issue/target/debug/deps/libcblas_sys-b604b97d9bc0dc00.rlib" "~/ndarray_blas_issue/target/debug/deps/liblibc-52f675a24655d4b8.rlib" "~/ndarray_blas_issue/target/debug/deps/libblas_src-0499c565dd169e87.rlib" "-Wl,--start-group" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-74488c47a41eb313.rlib" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-efbc2c947951b8cc.rlib" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-b177382d9a04ffbe.rlib" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-ba7d2244d33447d5.rlib" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libbacktrace-39569dc87e4ea301.rlib" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libbacktrace_sys-cd9e255bd82ddb5c.rlib" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-f83165e777dccb8d.rlib" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-10b085de263b1750.rlib" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-9a4a22edf44da957.rlib" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-68fabc677efa98de.rlib" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-026dc0061b48e8b6.rlib" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-7bf75bb619341145.rlib" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-ef54709e300503ed.rlib" "-Wl,--end-group" "~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-4ed27d8420cb4abc.rlib" "-Wl,-Bdynamic" "-lutil" "-lutil" "-ldl" "-lrt" "-lpthread" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-lutil"
  = note: ~/ndarray_blas_issue/target/debug/deps/libliba-362c1514609b1fc9.rlib(liba-362c1514609b1fc9.2v889vurran3pts.rcgu.o): In function `ndarray::linalg::impl_linalg::_$LT$impl$u20$ndarray..ArrayBase$LT$S$C$ndarray..dimension..dim..Dim$LT$$u5b$usize$u3b$$u20$_$u5d$$GT$$GT$$GT$::dot_impl::hb19ad3854e46ef64':
          ~/.cargo/registry/src/github.com-1ecc6299db9ec823/ndarray-0.13.0/src/linalg/impl_linalg.rs:114: undefined reference to `cblas_sdot'
          ~/.cargo/registry/src/github.com-1ecc6299db9ec823/ndarray-0.13.0/src/linalg/impl_linalg.rs:114: undefined reference to `cblas_ddot'
          collect2: error: ld returned 1 exit status
          

error: aborting due to previous error

error: could not compile `bina`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

Note that this also doesn't work if I make every crate depend on `ndarray-linalg`, which makes me think I didn't set it up correctly, but I think it is setup exactly like the `README` recommends.

**Comment by jturner314** [comment]
Thanks for creating a repository to illustrate the issue. That made it really easy to reproduce.

Here is an explanation of what's happening:

* If you run `cargo build` in `bin/bina`, `ndarray` is compiled without any features enabled.
* If you run `cargo build` at the top level of the workspace, `ndarray` is compiled with the `blas` feature enabled. (You can verify this by running `cargo build -v` and looking for the `rustc` call containing `--crate-name ndarray`.) When the `blas` feature is enabled and a binary is being built, `ndarray` requires a BLAS backend to be built and linked. Since no BLAS backend has been enabled in this example, the result is the linker error in your comment.

Why does including `libb` in the workspace cause this? Well, one of the differences of using a workspace versus individual crates outside of a workspace is that all crates in the workspace share a single `target` directory. When building a crate X of version Y in a workspace, `cargo` computes the union of all the features enabled for version Y of crate X  by all the crates in the workspace, and uses that set of features to build version Y of crate X only once. `libb` depends on `ndarray-linalg`, and `ndarray-linalg` enables the `blas` feature of its `ndarray` dependency. So, `ndarray` is built once with the `blas` feature enabled.

For the time being, there are a few workarounds:

* [build and link a BLAS backend](https://github.com/rust-ndarray/ndarray#how-to-use-with-cargo) in `bina` (this doesn't require `ndarray-linalg` or LAPACK)
* add a `ndarray-linalg` dependency to `bina` with one of the [backends enabled](https://github.com/rust-ndarray/ndarray-linalg#blaslapack-backend) (this is just another way to build/link a BLAS backend, but it also includes `ndarray-linalg` and a LAPACK backend)
* move the crate with the `ndarray-linalg` dependency into a separate workspace.

@termoshtt Why does `ndarray-linalg` enable the `blas` feature of `ndarray` and depend on `blas-src`? [If I remove them](https://github.com/rust-ndarray/ndarray-linalg/compare/master...jturner314:remove-blas?expand=1), the examples still build and run fine. Would it make sense to remove them?

**Comment by tobithiel** [comment]
Thanks for the quick response and detailed explanations.
I tried workaround 1 & 2 without much success. I'm still getting the same error, even when `liba`, `libb`, `bina` all have the same `ndarray` dependencies:
```
ndarray = { version = "0.13" }
ndarray-linalg = { version = "0.12", features = ["openblas"] }
```
Then everything should be the same, right?
I'm running `cargo clean && cargo build` to make sure no intermediate state is being left around.
I'm not sure if I'm missing something or misunderstanding something :confused:

**Comment by jturner314** [comment]
Oh, you're right. I should have tested my suggestions. :) I'll investigate further.

**Comment by jturner314** [comment]
The problem with my first suggestion was that the `ndarray` README listed the incorrect versions. I've created a PR to fix this (rust-ndarray/ndarray#777).

The problem with my second suggestion was that `ndarray` 0.13 and `ndarray-linalg` 0.12 depend on different versions of `blas-src`/`openblas-src`, so enabling the `openblas` feature of `ndarray-linalg` enabled one version of `openblas-src` (the one needed for `ndarray-linalg`) but not the version needed for `ndarray`.

Changing the `[dependencies]` section of `bin/bina/Cargo.toml` to the following fixes the linker issue:

```toml
[dependencies]
liba = { path = "../../lib/liba/" }
ndarray = { version = "0.13" }
blas-src = { version = "0.2.0", default-features = false, features = ["openblas"] }
openblas-src = { version = "0.6.0", default-features = false, features = ["cblas"] }
# ndarray-linalg = { version = "0.12", features = ["openblas"] }
```

**Comment by tobithiel** [comment]
Ah, that explains it! Thank so much for the super quick help.
